anndata/_io/merge.py

```diff
@@ -0,0 +1,647 @@
+        output_group[out_path].attrs.update(
+            {"encoding-type": "array", "encoding-version": "0.2.0"}
+        )
+
+
+def write_concat_sequence(
+    groups: Sequence[Union[ZarrGroup, H5Group]],
+    output_group,
+    out_path,
+    axis=0,
+    index=None,
+    reindexers=None,
+    fill_value=None,
+    join="inner",
+):
+    """
+    array, dataframe, csc_matrix, csc_matrix
+    """
+    arrays: Union[pd.DataFrame, SparseDataset, H5Array, ZarrArray] = [
+        read_group(g) for g in groups
+    ]
+    if any(isinstance(a, pd.DataFrame) for a in arrays):
+        if reindexers is None:
+            if join == "inner":
+                reindexers = gen_reindexers_df_inner(arrays, axis=axis)
+            else:
+                raise NotImplementedError("Cannot reindex dataframes with outer join.")
+        if not all(
+            isinstance(a, pd.DataFrame) or a is MissingVal or 0 in a.shape
+            for a in arrays
+        ):
+            raise NotImplementedError(
+                "Cannot concatenate a dataframe with other array types."
+            )
+        df = concat_arrays(
+            arrays=arrays,
+            reindexers=reindexers,
+            axis=axis,
+            index=index,
+            fill_value=fill_value,
+        )
+        write_elem(output_group, out_path, df)
+    # If all are compatible sparse matrices
+    elif all(
+        isinstance(a, (pd.DataFrame, SparseDataset, H5Array, ZarrArray)) for a in arrays
+    ):
+        write_concat_arrays(
+            arrays, output_group, out_path, axis, reindexers, fill_value, join
+        )
+    else:
+        raise NotImplementedError(
+            f"Concatenation of these types is not yet implemented: {[get_encoding_type(g) for g in groups] } with axis={axis}."
+        )
+
+
+def _get_group_from_hdf5(store, *args, **kwargs) -> H5Group:
+    import h5py
+
+    return h5py.File(store, *args, **kwargs)
+
+
+@singledispatch
+def _get_group(store, *args, **kwargs) -> Union[ZarrGroup, H5Group]:
+    raise NotImplementedError("This is not yet implemented.")
+
+
+@_get_group.register
+def _(store: Path, *args, **kwargs) -> Union[ZarrGroup, H5Group]:
+    if store.suffix == ".h5ad":
+        return _get_group_from_hdf5(store, *args, **kwargs)
+    import zarr
+
+    return zarr.open_group(store, *args, **kwargs)
+
+
+@_get_group.register
+def _(store: str, *args, **kwargs) -> Union[ZarrGroup, H5Group]:
+    if store.endswith(".h5ad"):
+        return _get_group_from_hdf5(store, *args, **kwargs)
+    import zarr
+
+    return zarr.open_group(store, *args, **kwargs)
+
+
+@_get_group.register
+def _(store: dict, *args, **kwargs) -> ZarrGroup:
+    import zarr
+
+    return zarr.open_group(store, *args, **kwargs)
+
+
+@_get_group.register
+def _(store: ZarrGroup, *args, **kwargs) -> ZarrGroup:
+    return store
+
+
+@_get_group.register
+def _(store: H5Group, *args, **kwargs) -> H5Group:
+    return store
+
+
+def _get_groups_from_paths(
```

Comment:
I think this may be simple enough to just go inline.

Otherwise I think it should at least be renamed, like `_get_group`.